### PR TITLE
[RISCV] Split vector FP pseudo instructions by SEW

### PR DESCRIPTION
**Pull request by michaelmaitland**
This patch contains a series of commits to split vector FP pseudo instructions by SEW. This is intended to be the equivalent of https://reviews.llvm.org/D153000. I will add @wangpc-pp as co-author. Merging this PR closes https://github.com/llvm/llvm-project/pull/87404. Scheduler changes depending on SEW can come in future patches.

There are still some FP instructions that can be split, but they were not split in https://reviews.llvm.org/D153000 so I did not split them here. We can split those in a follow up patch or in this one if we'd like.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td [hunk 1]
```diff
@@ -2823,16 +2831,17 @@ multiclass VPseudoVDIV_VV_VX {
 
 multiclass VPseudoVFMUL_VV_VF_RM {
   foreach m = MxListF in {
-    defm "" : VPseudoBinaryFV_VV_RM<m>,
-              SchedBinary<"WriteVFMulV", "ReadVFMulV", "ReadVFMulV", m.MX,
-                          forceMergeOpRead=true>;
+    foreach e = SchedSEWSet<m.MX, isF=1>.val in
+      defm "" : VPseudoBinaryFV_VV_RM<m, "", e>,
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td:2835]: We can use named argument to pass `sew` here.

File: llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td [hunk 2]
```diff
@@ -2823,16 +2831,17 @@ multiclass VPseudoVDIV_VV_VX {
...
+                SchedBinary<"WriteVFMulV", "ReadVFMulV", "ReadVFMulV", m.MX, e,
+                            forceMergeOpRead=true>;
   }
 
   foreach f = FPList in {
     foreach m = f.MxList in {
-      defm "" : VPseudoBinaryV_VF_RM<m, f>,
+      defm "" : VPseudoBinaryV_VF_RM<m, f, "", f.SEW>,
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td:2842]: ditto.

File: llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td [hunk 3]
```diff
@@ -2912,26 +2921,27 @@ multiclass VPseudoVMAX_VV_VF {
 
 multiclass VPseudoVALU_VV_VF_RM {
   foreach m = MxListF in {
-    defm "" : VPseudoBinaryFV_VV_RM<m>,
-              SchedBinary<"WriteVFALUV", "ReadVFALUV", "ReadVFALUV", m.MX,
-                          forceMergeOpRead=true>;
+    foreach e = SchedSEWSet<m.MX, isF=1>.val in
+      defm "" : VPseudoBinaryFV_VV_RM<m, "", e>,
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td:2925]: ditto.

File: llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td [hunk 4]
```diff
@@ -2912,26 +2921,27 @@ multiclass VPseudoVMAX_VV_VF {
...
+                SchedBinary<"WriteVFALUV", "ReadVFALUV", "ReadVFALUV", m.MX, e,
+                            forceMergeOpRead=true>;
   }
 
   foreach f = FPList in {
     foreach m = f.MxList in {
-      defm "" : VPseudoBinaryV_VF_RM<m, f>,
+      defm "" : VPseudoBinaryV_VF_RM<m, f, "", f.SEW>,
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td:2932]: ditto.

File: llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td [hunk 5]
```diff
@@ -2912,26 +2921,27 @@ multiclass VPseudoVMAX_VV_VF {
...
+                            f.SEW, forceMergeOpRead=true>;
     }
   }
 }
 
 multiclass VPseudoVALU_VF_RM {
   foreach f = FPList in {
     foreach m = f.MxList in {
-      defm "" : VPseudoBinaryV_VF_RM<m, f>,
+      defm "" : VPseudoBinaryV_VF_RM<m, f, "", f.SEW>,
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td:2942]: ditto.

File: llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td [hunk 6]
```diff
@@ -5693,7 +5742,7 @@ multiclass VPatTernaryW_VX<string intrinsic, string instruction,
 }
 
 multiclass VPatTernaryW_VX_RM<string intrinsic, string instruction,
-                           list<VTypeInfoToWide> vtilist> {
+                           list<VTypeInfoToWide> vtilist, bit isSEWAware = 0> {
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td:5745]: Fix the indent by the way. :-)

**Review by wangpc-pp — COMMENTED** [review]
Thanks! I don't have correctness concerns here but I need a second eye.
And, can you provide some data about how many bytes the MatcherTable will increase?

**Comment by michaelmaitland** [comment]
> Thanks! I don't have correctness concerns here but I need a second eye. And, can you provide some data about how many bytes the MatcherTable will increase?

I am seeing the MatcherTable size stay the same at 2216191 bytes.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 7]
```diff
@@ -2603,31 +2603,54 @@ std::string RISCVInstrInfo::createMIROperandComment(
...
+#define CASE_VFMA_OPCODE_LMULS_MF2(OP, TYPE, SEW)                              \
+  CASE_VFMA_OPCODE_COMMON(OP, TYPE, MF2, SEW):                                 \
+  case CASE_VFMA_OPCODE_LMULS_M1(OP, TYPE, SEW)
+
+#define CASE_VFMA_OPCODE_LMULS_MF4(OP, TYPE, SEW)                              \
+  CASE_VFMA_OPCODE_COMMON(OP, TYPE, MF4, SEW):                                 \
+  case CASE_VFMA_OPCODE_LMULS_MF2(OP, TYPE, SEW)
+
+#define CASE_VFMA_OPCODE_VV(OP)                                                \
+  CASE_VFMA_OPCODE_COMMON(OP, VV, MF4, E16):                                   \
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:2646]: Can the first two lines of these becombined to `CASE_VFMA_OPCODE_LMULS_MF4(OP, VV, E16)`?

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp]: Updated

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp]: Are we missing test coverage for E64?

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp]: Did failures exist yesterday when I made the original comment?

**Review by wangpc-pp — APPROVED** [review]
LGTM.

**Comment by michaelmaitland** [comment]
This PR is closed by the individual commits 5fa58e28f34f4b5b41dd2869212ac917fb42b2e3, aece68269c8cb28e16dc5c4ec9aabecb807a0438, d309d7e4fe51a0ff0ce6430a8af585e2af2772a3, c6b7944be4dfbb1fb35301c670812726845acaa7, 43248ffea7b8de3a33b11768e8c21d2434252528, andb8e1ff322104e5445371328a4fbc9c1d2bf82f01

**Pull request closed without merging**